// usb/dist/usb/device.js
"use strict";Object.defineProperty(exports,"__esModule",{value:!0}),exports.ExtendedDevice=void 0;var usb=require("./bindings"),interface_1=require("./interface"),capability_1=require("./capability"),isBuffer=function(e){return!!e&&e instanceof Uint8Array},DEFAULT_TIMEOUT=1e3,ExtendedDevice=function(){function e(){this._timeout=DEFAULT_TIMEOUT}return Object.defineProperty(e.prototype,"timeout",{get:function(){return this._timeout||DEFAULT_TIMEOUT},set:function(e){this._timeout=e},enumerable:!1,configurable:!0}),Object.defineProperty(e.prototype,"configDescriptor",{get:function(){try{return this.__getConfigDescriptor()}catch(e){if(e.errno===usb.LIBUSB_ERROR_NOT_FOUND)return;throw e}},enumerable:!1,configurable:!0}),Object.defineProperty(e.prototype,"allConfigDescriptors",{get:function(){try{return this.__getAllConfigDescriptors()}catch(e){if(e.errno===usb.LIBUSB_ERROR_NOT_FOUND)return[];throw e}},enumerable:!1,configurable:!0}),Object.defineProperty(e.prototype,"parent",{get:function(){return this.__getParent()},enumerable:!1,configurable:!0}),e.prototype.open=function(e){if(void 0===e&&(e=!0),this.__open(),!1!==e){this.interfaces=[];for(var t=this.configDescriptor?this.configDescriptor.interfaces.length:0,r=0;r<t;r++)this.interfaces[r]=new interface_1.Interface(this,r)}},e.prototype.close=function(){this.__close(),this.interfaces=void 0},e.prototype.setConfiguration=function(e,t){var r=this;this.__setConfiguration(e,(function(e){if(!e){r.interfaces=[];for(var i=r.configDescriptor?r.configDescriptor.interfaces.length:0,n=0;n<i;n++)r.interfaces[n]=new interface_1.Interface(r,n)}t&&t.call(r,e)}))},e.prototype.controlTransfer=function(e,t,r,i,n,o){var s=this,c=!!(e&usb.LIBUSB_ENDPOINT_IN),f=c?n:n.length;if(c){if(!(n>=0))throw new TypeError("Expected size number for IN transfer (based on bmRequestType)")}else if(!isBuffer(n))throw new TypeError("Expected buffer for OUT transfer (based on bmRequestType)");var a=Buffer.alloc(f+usb.LIBUSB_CONTROL_SETUP_SIZE);a.writeUInt8(e,0),a.writeUInt8(t,1),a.writeUInt16LE(r,2),a.writeUInt16LE(i,4),a.writeUInt16LE(f,6),c||a.set(n,usb.LIBUSB_CONTROL_SETUP_SIZE);var u=new usb.Transfer(this,0,usb.LIBUSB_TRANSFER_TYPE_CONTROL,this.timeout,(function(e,t,r){o&&(c?o.call(s,e,t.slice(usb.LIBUSB_CONTROL_SETUP_SIZE,usb.LIBUSB_CONTROL_SETUP_SIZE+r)):o.call(s,e,r))}));try{u.submit(a)}catch(e){o&&process.nextTick((function(){return o.call(s,e,void 0)}))}return this},e.prototype.interface=function(e){if(!this.interfaces)throw new Error("Device must be open before searching for interfaces");e=e||0;for(var t=0;t<this.interfaces.length;t++)if(this.interfaces[t].interfaceNumber===e)return this.interfaces[t];throw new Error("Interface not found for address: "+e)},e.prototype.getStringDescriptor=function(e,t){0!==e?this.controlTransfer(usb.LIBUSB_ENDPOINT_IN,usb.LIBUSB_REQUEST_GET_DESCRIPTOR,usb.LIBUSB_DT_STRING<<8|e,1033,255,(function(e,r){if(e)return t(e);t(void 0,isBuffer(r)?r.toString("utf16le",2):void 0)})):t()},e.prototype.getBosDescriptor=function(e){var t=this;return this._bosDescriptor?e(void 0,this._bosDescriptor):this.deviceDescriptor.bcdUSB<513?e(void 0,void 0):void this.controlTransfer(usb.LIBUSB_ENDPOINT_IN,usb.LIBUSB_REQUEST_GET_DESCRIPTOR,usb.LIBUSB_DT_BOS<<8,0,usb.LIBUSB_DT_BOS_SIZE,(function(r,i){if(r)return r.errno===usb.LIBUSB_TRANSFER_STALL?e(void 0,void 0):e(r,void 0);if(!isBuffer(i))return e(void 0,void 0);var n=i.readUInt16LE(2);t.controlTransfer(usb.LIBUSB_ENDPOINT_IN,usb.LIBUSB_REQUEST_GET_DESCRIPTOR,usb.LIBUSB_DT_BOS<<8,0,n,(function(r,i){if(r)return r.errno===usb.LIBUSB_TRANSFER_STALL?e(void 0,void 0):e(r,void 0);if(!isBuffer(i))return e(void 0,void 0);for(var n={bLength:i.readUInt8(0),bDescriptorType:i.readUInt8(1),wTotalLength:i.readUInt16LE(2),bNumDeviceCaps:i.readUInt8(4),capabilities:[]},o=usb.LIBUSB_DT_BOS_SIZE;o<n.wTotalLength;){var s={bLength:i.readUInt8(o+0),bDescriptorType:i.readUInt8(o+1),bDevCapabilityType:i.readUInt8(o+2),dev_capability_data:i.slice(o+3,o+i.readUInt8(o+0))};n.capabilities.push(s),o+=s.bLength}t._bosDescriptor=n,e(void 0,t._bosDescriptor)}))}))},e.prototype.getCapabilities=function(e){var t=this,r=[];this.getBosDescriptor((function(i,n){if(i)return e(i,void 0);for(var o=n?n.capabilities.length:0,s=0;s<o;s++)r.push(new capability_1.Capability(t,s));e(void 0,r)}))},e}();exports.ExtendedDevice=ExtendedDevice;
// SIG // Begin signature block
// SIG // MIIoQAYJKoZIhvcNAQcCoIIoMTCCKC0CAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // oQnhoZcB97ZBq3SbwRSk6ZGdLPGslWVKnhAUxeq29Tig
// SIG // gg2LMIIGCTCCA/GgAwIBAgITMwAAA4TZaH1mzHVLoQAA
// SIG // AAADhDANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTIzMDcxMzIzNDUzOFoX
// SIG // DTI0MDkxNTIzNDUzOFowgYgxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xMjAwBgNVBAMTKU1pY3Jvc29mdCAzcmQgUGFydHkg
// SIG // QXBwbGljYXRpb24gQ29tcG9uZW50MIIBIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAQ8AMIIBCgKCAQEAznLqoz/10YOqShXj
// SIG // Gln6TE4BGUwZ3uJsXV3BIhK2/5WfWH8YMWvFoTgOul1P
// SIG // rrDlA5qOs/1qSNylRSLn+DKx4e+sQC/x8qVPWx+4bwhu
// SIG // NHaNGEMZsppPjtuc10OmdECsioQzVOyu/TlVohP0DUT1
// SIG // w8V4+uR06k9qm+XVcw84nrH6xac865B7TswYpBBBZdaG
// SIG // X5mTst5UnoomlS8JIGqD6WDAyvlmwDRl0C0Omj+b8lg7
// SIG // ZJ4wz9hfYI6eo/kOu2wgYyLPBaXR9iDehwLxN1/xOGDi
// SIG // Dq2WNiWFM/aaBYda645BrjOAk9rc3aoX5AdvEDp7oRMi
// SIG // fgJgdwdNaC/WoTUomQIDAQABo4IBczCCAW8wHwYDVR0l
// SIG // BBgwFgYKKwYBBAGCN0wRAQYIKwYBBQUHAwMwHQYDVR0O
// SIG // BBYEFBzcFkQMR/V8OGV/cF0ZRXO9ery8MEUGA1UdEQQ+
// SIG // MDykOjA4MR4wHAYDVQQLExVNaWNyb3NvZnQgQ29ycG9y
// SIG // YXRpb24xFjAUBgNVBAUTDTIzMTUyMis1MDExNDgwHwYD
// SIG // VR0jBBgwFoAUSG5k5VAF04KqFzc3IrVtqMp1ApUwVAYD
// SIG // VR0fBE0wSzBJoEegRYZDaHR0cDovL3d3dy5taWNyb3Nv
// SIG // ZnQuY29tL3BraW9wcy9jcmwvTWljQ29kU2lnUENBMjAx
// SIG // MV8yMDExLTA3LTA4LmNybDBhBggrBgEFBQcBAQRVMFMw
// SIG // UQYIKwYBBQUHMAKGRWh0dHA6Ly93d3cubWljcm9zb2Z0
// SIG // LmNvbS9wa2lvcHMvY2VydHMvTWljQ29kU2lnUENBMjAx
// SIG // MV8yMDExLTA3LTA4LmNydDAMBgNVHRMBAf8EAjAAMA0G
// SIG // CSqGSIb3DQEBCwUAA4ICAQAnvqLcbzrt8+kzifeRGz6b
// SIG // UaHMp1aMqJtfMarjiUinqgzSLbtqZslaWomQ4T7LS8rq
// SIG // NJl8jSz4d0wJfm6LblH6RWQzhs4JqllQa3R0MbLt7fvs
// SIG // ehsNASX0L5/3NP5VP40U0UwKe0u4wKo6qcjHNMr66y+s
// SIG // b1DZ965C5Hctslwx1qR3RMeSXaSQ6YJ+Y9bb9xhnbXmH
// SIG // YWZW9h6JWnFYriUm3jHRubAw7KXMPMWNQ7OBTF6uzGKi
// SIG // 2ELK/CsFyWfeUm6GdPVf+TDsKShhoZXtAXOZwBK3SQTY
// SIG // 42gSwTBg4oPTldGd2SH2OIzhZXkxLCuSzk0LUmjVQ+K2
// SIG // xvkJzLMHi2oO6lTO8+SHjqhO1GPwnwhJLJfoaynNzAb3
// SIG // XMj3CdvPcd4EZDh9PRSI0SAHxaQQ0pD4+B1uSyqjUr9w
// SIG // GxZbisemYGzWIcB8UEsQ6mLE5QI7Z2zI2CNkU60iDI4B
// SIG // mEb38Gue8qcj3Zcs5teEOVDYILFsq9ToPkG0HluvkQhu
// SIG // hblf5xAGKzIHh5TEAXedGlYU3QXquQkdxs+quGlS+5xH
// SIG // Q0enmLi9FNXuBKC9jmAaJN8uKwS96+e5yxBGesTBgHuX
// SIG // hwT7YHLj8OUv2iQ66KXrvuAV0ouUf6hpLU5Gyn5STAJX
// SIG // 1xy+muuKPPIfIvGWeBxxzONasTAedo/W4GggkBSRZSyT
// SIG // 7zCCB3owggVioAMCAQICCmEOkNIAAAAAAAMwDQYJKoZI
// SIG // hvcNAQELBQAwgYgxCzAJBgNVBAYTAlVTMRMwEQYDVQQI
// SIG // EwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4w
// SIG // HAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xMjAw
// SIG // BgNVBAMTKU1pY3Jvc29mdCBSb290IENlcnRpZmljYXRl
// SIG // IEF1dGhvcml0eSAyMDExMB4XDTExMDcwODIwNTkwOVoX
// SIG // DTI2MDcwODIxMDkwOVowfjELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEoMCYGA1UEAxMfTWljcm9zb2Z0IENvZGUgU2lnbmlu
// SIG // ZyBQQ0EgMjAxMTCCAiIwDQYJKoZIhvcNAQEBBQADggIP
// SIG // ADCCAgoCggIBAKvw+nIQHC6t2G6qghBNNLrytlghn0Ib
// SIG // KmvpWlCquAY4GgRJun/DDB7dN2vGEtgL8DjCmQawyDnV
// SIG // ARQxQtOJDXlkh36UYCRsr55JnOloXtLfm1OyCizDr9mp
// SIG // K656Ca/XllnKYBoF6WZ26DJSJhIv56sIUM+zRLdd2MQu
// SIG // A3WraPPLbfM6XKEW9Ea64DhkrG5kNXimoGMPLdNAk/jj
// SIG // 3gcN1Vx5pUkp5w2+oBN3vpQ97/vjK1oQH01WKKJ6cuAS
// SIG // OrdJXtjt7UORg9l7snuGG9k+sYxd6IlPhBryoS9Z5JA7
// SIG // La4zWMW3Pv4y07MDPbGyr5I4ftKdgCz1TlaRITUlwzlu
// SIG // ZH9TupwPrRkjhMv0ugOGjfdf8NBSv4yUh7zAIXQlXxgo
// SIG // tswnKDglmDlKNs98sZKuHCOnqWbsYR9q4ShJnV+I4iVd
// SIG // 0yFLPlLEtVc/JAPw0XpbL9Uj43BdD1FGd7P4AOG8rAKC
// SIG // X9vAFbO9G9RVS+c5oQ/pI0m8GLhEfEXkwcNyeuBy5yTf
// SIG // v0aZxe/CHFfbg43sTUkwp6uO3+xbn6/83bBm4sGXgXvt
// SIG // 1u1L50kppxMopqd9Z4DmimJ4X7IvhNdXnFy/dygo8e1t
// SIG // wyiPLI9AN0/B4YVEicQJTMXUpUMvdJX3bvh4IFgsE11g
// SIG // lZo+TzOE2rCIF96eTvSWsLxGoGyY0uDWiIwLAgMBAAGj
// SIG // ggHtMIIB6TAQBgkrBgEEAYI3FQEEAwIBADAdBgNVHQ4E
// SIG // FgQUSG5k5VAF04KqFzc3IrVtqMp1ApUwGQYJKwYBBAGC
// SIG // NxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8G
// SIG // A1UdEwEB/wQFMAMBAf8wHwYDVR0jBBgwFoAUci06AjGQ
// SIG // Q7kUBU7h6qfHMdEjiTQwWgYDVR0fBFMwUTBPoE2gS4ZJ
// SIG // aHR0cDovL2NybC5taWNyb3NvZnQuY29tL3BraS9jcmwv
// SIG // cHJvZHVjdHMvTWljUm9vQ2VyQXV0MjAxMV8yMDExXzAz
// SIG // XzIyLmNybDBeBggrBgEFBQcBAQRSMFAwTgYIKwYBBQUH
// SIG // MAKGQmh0dHA6Ly93d3cubWljcm9zb2Z0LmNvbS9wa2kv
// SIG // Y2VydHMvTWljUm9vQ2VyQXV0MjAxMV8yMDExXzAzXzIy
// SIG // LmNydDCBnwYDVR0gBIGXMIGUMIGRBgkrBgEEAYI3LgMw
// SIG // gYMwPwYIKwYBBQUHAgEWM2h0dHA6Ly93d3cubWljcm9z
// SIG // b2Z0LmNvbS9wa2lvcHMvZG9jcy9wcmltYXJ5Y3BzLmh0
// SIG // bTBABggrBgEFBQcCAjA0HjIgHQBMAGUAZwBhAGwAXwBw
// SIG // AG8AbABpAGMAeQBfAHMAdABhAHQAZQBtAGUAbgB0AC4g
// SIG // HTANBgkqhkiG9w0BAQsFAAOCAgEAZ/KGpZjgVHkaLtPY
// SIG // dGcimwuWEeFjkplCln3SeQyQwWVfLiw++MNy0W2D/r4/
// SIG // 6ArKO79HqaPzadtjvyI1pZddZYSQfYtGUFXYDJJ80hpL
// SIG // HPM8QotS0LD9a+M+By4pm+Y9G6XUtR13lDni6WTJRD14
// SIG // eiPzE32mkHSDjfTLJgJGKsKKELukqQUMm+1o+mgulaAq
// SIG // PyprWEljHwlpblqYluSD9MCP80Yr3vw70L01724lruWv
// SIG // J+3Q3fMOr5kol5hNDj0L8giJ1h/DMhji8MUtzluetEk5
// SIG // CsYKwsatruWy2dsViFFFWDgycScaf7H0J/jeLDogaZiy
// SIG // WYlobm+nt3TDQAUGpgEqKD6CPxNNZgvAs0314Y9/HG8V
// SIG // fUWnduVAKmWjw11SYobDHWM2l4bf2vP48hahmifhzaWX
// SIG // 0O5dY0HjWwechz4GdwbRBrF1HxS+YWG18NzGGwS+30HH
// SIG // Diju3mUv7Jf2oVyW2ADWoUa9WfOXpQlLSBCZgB/QACnF
// SIG // sZulP0V3HjXG0qKin3p6IvpIlR+r+0cjgPWe+L9rt0uX
// SIG // 4ut1eBrs6jeZeRhL/9azI2h15q/6/IvrC4DqaTuv/DDt
// SIG // BEyO3991bWORPdGdVk5Pv4BXIqF4ETIheu9BCrE/+6jM
// SIG // pF3BoYibV3FWTkhFwELJm3ZbCoBIa/15n8G9bW1qyVJz
// SIG // Ew16UM0xghoNMIIaCQIBATCBlTB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExAhMzAAADhNlofWbMdUuhAAAA
// SIG // AAOEMA0GCWCGSAFlAwQCAQUAoIGuMBkGCSqGSIb3DQEJ
// SIG // AzEMBgorBgEEAYI3AgEEMBwGCisGAQQBgjcCAQsxDjAM
// SIG // BgorBgEEAYI3AgEVMC8GCSqGSIb3DQEJBDEiBCCI9Ub6
// SIG // mH1gx+SI4kCtWqlTcxJssWHK4oMGCtrIB2ANMDBCBgor
// SIG // BgEEAYI3AgEMMTQwMqAUgBIATQBpAGMAcgBvAHMAbwBm
// SIG // AHShGoAYaHR0cDovL3d3dy5taWNyb3NvZnQuY29tMA0G
// SIG // CSqGSIb3DQEBAQUABIIBADUHwiVLXyK1M0QHriguTifo
// SIG // IGOYk74ut8gt/qhx8l9yHHFlGg606Rm1JLlCpAtCc2xa
// SIG // 66KomZFfFpcETymhYjun5htaISFY4rlT0yIj3kPMHa2u
// SIG // CYrXg5eJsu5IkIan/rwIy3AqUUD+wGjAapnZGofpiyM/
// SIG // AATmYCBRgVVoY0oJlWgoTXBTEgV4DVCDQKVFJuuGVfU+
// SIG // /mD/4IVlTD4clIYGUSa4uh6OI0Ts9A+rzvAmJpCiCHUc
// SIG // tzJ6H8T9m4Ic1pEyPCecYzyHU7C2jl3TeyXJscn17vWu
// SIG // bIVn9jbnm/U0U9KzDNzvqbyDNFIyvkq7URswRdHM9CsU
// SIG // Y4w3r0Hs/wKhgheXMIIXkwYKKwYBBAGCNwMDATGCF4Mw
// SIG // ghd/BgkqhkiG9w0BBwKgghdwMIIXbAIBAzEPMA0GCWCG
// SIG // SAFlAwQCAQUAMIIBUgYLKoZIhvcNAQkQAQSgggFBBIIB
// SIG // PTCCATkCAQEGCisGAQQBhFkKAwEwMTANBglghkgBZQME
// SIG // AgEFAAQgNvNQ12g4vxOiL8BpyN+POOeIzjMvRGKfNuYR
// SIG // 62cvMtkCBmV7C/NTkxgTMjAyNDAxMDYwMjE5MTIuMDQ4
// SIG // WjAEgAIB9KCB0aSBzjCByzELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjElMCMGA1UECxMcTWljcm9zb2Z0IEFtZXJpY2EgT3Bl
// SIG // cmF0aW9uczEnMCUGA1UECxMeblNoaWVsZCBUU1MgRVNO
// SIG // OjdGMDAtMDVFMC1EOTQ3MSUwIwYDVQQDExxNaWNyb3Nv
// SIG // ZnQgVGltZS1TdGFtcCBTZXJ2aWNloIIR7TCCByAwggUI
// SIG // oAMCAQICEzMAAAHVqQLPxafJ6VoAAQAAAdUwDQYJKoZI
// SIG // hvcNAQELBQAwfDELMAkGA1UEBhMCVVMxEzARBgNVBAgT
// SIG // Cldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAc
// SIG // BgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQG
// SIG // A1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIw
// SIG // MTAwHhcNMjMwNTI1MTkxMjMwWhcNMjQwMjAxMTkxMjMw
// SIG // WjCByzELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hp
// SIG // bmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoT
// SIG // FU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjElMCMGA1UECxMc
// SIG // TWljcm9zb2Z0IEFtZXJpY2EgT3BlcmF0aW9uczEnMCUG
// SIG // A1UECxMeblNoaWVsZCBUU1MgRVNOOjdGMDAtMDVFMC1E
// SIG // OTQ3MSUwIwYDVQQDExxNaWNyb3NvZnQgVGltZS1TdGFt
// SIG // cCBTZXJ2aWNlMIICIjANBgkqhkiG9w0BAQEFAAOCAg8A
// SIG // MIICCgKCAgEAxX2pOezqYfb7sbZaAAYi3Onp0/sih+tf
// SIG // W/joRnZMoYDc5F/NClBiP4xKjlTFeEqrf1DxRYncdre7
// SIG // 9khE49rQM7lSrQ36thwabvNL2dL8kA8nVbeDAy+LSUiq
// SIG // oGKHwsQyAa1sySY4AaJSnTicdbrdk8jnPlDpu3vdrVTx
// SIG // 6Y3YPjpn99Uy1H/6KLsgDifyJ59oodwEj9EGJvgBUI4W
// SIG // AzQ7vLszHcBxeUHcLLHDWvT1UhnnS3Qy6PYy+g6DxeDW
// SIG // KNOpb7xELkifSJsGXwRi8v/IaRO0Q+HsLySpjNfrenkL
// SIG // hLE146xjNlo5FtfEoFGfJ/laS9rpOgIQ5Amt+eSOd9ad
// SIG // CZKqaKJ+3R7i1HWUkDuNKplSEOqkAmp7yJk6pjYBP6zy
// SIG // dK4K9ITDyP7kdU/4mi9JhKuG6mpPt7GvCPhQGDiPzwu1
// SIG // fsxHpPrHclrWl/p3Wxpb/0SW+ZkGhp/Dbp25H7xw9ULe
// SIG // Q9K5rTDnpGDKu0I2KhNxRD/8AGOEw7icbLY7Gth14tsl
// SIG // AxIODCS+vyb7EF06DmfiMUeik+beXweRaWWAaVSzJmt6
// SIG // Zuc+lV75F62LN9zVyalVi8IrMGuzVBVfOxLNrKLvSHcN
// SIG // 8gGZhyGFoDkPgyGNZ2N2huQzg4sDdaychG/pm1g2oK3V
// SIG // cXJ3K+lCijuPXqDU1xFvrHFH+hsFxtChMpkCAwEAAaOC
// SIG // AUkwggFFMB0GA1UdDgQWBBSv0FyjTt+WwTDs90eYIl+w
// SIG // JWtjmjAfBgNVHSMEGDAWgBSfpxVdAF5iXYP05dJlpxtT
// SIG // NRnpcjBfBgNVHR8EWDBWMFSgUqBQhk5odHRwOi8vd3d3
// SIG // Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2NybC9NaWNyb3Nv
// SIG // ZnQlMjBUaW1lLVN0YW1wJTIwUENBJTIwMjAxMCgxKS5j
// SIG // cmwwbAYIKwYBBQUHAQEEYDBeMFwGCCsGAQUFBzAChlBo
// SIG // dHRwOi8vd3d3Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2Nl
// SIG // cnRzL01pY3Jvc29mdCUyMFRpbWUtU3RhbXAlMjBQQ0El
// SIG // MjAyMDEwKDEpLmNydDAMBgNVHRMBAf8EAjAAMBYGA1Ud
// SIG // JQEB/wQMMAoGCCsGAQUFBwMIMA4GA1UdDwEB/wQEAwIH
// SIG // gDANBgkqhkiG9w0BAQsFAAOCAgEAZkXXar0Lso0acdBx
// SIG // sbzqn3xa24BQINI/Nr9L+0fv/Gu9EtleLe8imCkrEBiO
// SIG // qb7dQPCRprmco5iF5sNtwJCOW2ZMb0oMLPK6wbScPb2p
// SIG // BHJHFhUG+yRum+tfMdLPUNwy9HUauOkGUZ5u1Ott+JXn
// SIG // L47LQKMN9HT9E5yGnD1iqT3N0IAflg54JTdn3U9a7kOl
// SIG // iFQXp5qY6RvcqUQDSlMeTUXKXSQEFthahB00tzbW/tLQ
// SIG // qiJDRyeWhbBenoUaL1madDGCM/W6SR4sdFa43S1TDqXu
// SIG // 8L+trfdBN1KxNiplcKUOcLDA+mFLHKArEkUGawOQG8Ez
// SIG // gmSaXhts97w6P4brzyvE3kydi7bvyLV4MSJSDkKf7Wxw
// SIG // IjfF6bcPyZiGYPXnUPxYg2iCMBuWB7H5tru08Dhcnejq
// SIG // i8NfGeY/yLwz85ZMFicZOkRyReXUuLN358i8NwxwXuQ2
// SIG // r+imAeJ/Mf3BJg/0eOP/IEuT37htbK4y3cshrcodCokQ
// SIG // 0po8Pn2u4tVT6HponQ1jWe5LDWnTGneGaA74JKjleAOm
// SIG // vjfByHPz+oNIq63sy1lIGyl0jfIh/UT/liRceXwRxOAB
// SIG // ca2wFENmZ+Yw5hwCN8GHEA55xGD+dQO+VhcD7Lwa3629
// SIG // fumtX7kxB9QGHTjjMvH1/MSqNBNGLPu28SLxT7FKUs3x
// SIG // YwaJZocwggdxMIIFWaADAgECAhMzAAAAFcXna54Cm0mZ
// SIG // AAAAAAAVMA0GCSqGSIb3DQEBCwUAMIGIMQswCQYDVQQG
// SIG // EwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UE
// SIG // BxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENv
// SIG // cnBvcmF0aW9uMTIwMAYDVQQDEylNaWNyb3NvZnQgUm9v
// SIG // dCBDZXJ0aWZpY2F0ZSBBdXRob3JpdHkgMjAxMDAeFw0y
// SIG // MTA5MzAxODIyMjVaFw0zMDA5MzAxODMyMjVaMHwxCzAJ
// SIG // BgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAw
// SIG // DgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3Nv
// SIG // ZnQgQ29ycG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jvc29m
// SIG // dCBUaW1lLVN0YW1wIFBDQSAyMDEwMIICIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAg8AMIICCgKCAgEA5OGmTOe0ciELeaLL
// SIG // 1yR5vQ7VgtP97pwHB9KpbE51yMo1V/YBf2xK4OK9uT4X
// SIG // YDP/XE/HZveVU3Fa4n5KWv64NmeFRiMMtY0Tz3cywBAY
// SIG // 6GB9alKDRLemjkZrBxTzxXb1hlDcwUTIcVxRMTegCjhu
// SIG // je3XD9gmU3w5YQJ6xKr9cmmvHaus9ja+NSZk2pg7uhp7
// SIG // M62AW36MEBydUv626GIl3GoPz130/o5Tz9bshVZN7928
// SIG // jaTjkY+yOSxRnOlwaQ3KNi1wjjHINSi947SHJMPgyY9+
// SIG // tVSP3PoFVZhtaDuaRr3tpK56KTesy+uDRedGbsoy1cCG
// SIG // MFxPLOJiss254o2I5JasAUq7vnGpF1tnYN74kpEeHT39
// SIG // IM9zfUGaRnXNxF803RKJ1v2lIH1+/NmeRd+2ci/bfV+A
// SIG // utuqfjbsNkz2K26oElHovwUDo9Fzpk03dJQcNIIP8BDy
// SIG // t0cY7afomXw/TNuvXsLz1dhzPUNOwTM5TI4CvEJoLhDq
// SIG // hFFG4tG9ahhaYQFzymeiXtcodgLiMxhy16cg8ML6EgrX
// SIG // Y28MyTZki1ugpoMhXV8wdJGUlNi5UPkLiWHzNgY1GIRH
// SIG // 29wb0f2y1BzFa/ZcUlFdEtsluq9QBXpsxREdcu+N+VLE
// SIG // hReTwDwV2xo3xwgVGD94q0W29R6HXtqPnhZyacaue7e3
// SIG // PmriLq0CAwEAAaOCAd0wggHZMBIGCSsGAQQBgjcVAQQF
// SIG // AgMBAAEwIwYJKwYBBAGCNxUCBBYEFCqnUv5kxJq+gpE8
// SIG // RjUpzxD/LwTuMB0GA1UdDgQWBBSfpxVdAF5iXYP05dJl
// SIG // pxtTNRnpcjBcBgNVHSAEVTBTMFEGDCsGAQQBgjdMg30B
// SIG // ATBBMD8GCCsGAQUFBwIBFjNodHRwOi8vd3d3Lm1pY3Jv
// SIG // c29mdC5jb20vcGtpb3BzL0RvY3MvUmVwb3NpdG9yeS5o
// SIG // dG0wEwYDVR0lBAwwCgYIKwYBBQUHAwgwGQYJKwYBBAGC
// SIG // NxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8G
// SIG // A1UdEwEB/wQFMAMBAf8wHwYDVR0jBBgwFoAU1fZWy4/o
// SIG // olxiaNE9lJBb186aGMQwVgYDVR0fBE8wTTBLoEmgR4ZF
// SIG // aHR0cDovL2NybC5taWNyb3NvZnQuY29tL3BraS9jcmwv
// SIG // cHJvZHVjdHMvTWljUm9vQ2VyQXV0XzIwMTAtMDYtMjMu
// SIG // Y3JsMFoGCCsGAQUFBwEBBE4wTDBKBggrBgEFBQcwAoY+
// SIG // aHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3BraS9jZXJ0
// SIG // cy9NaWNSb29DZXJBdXRfMjAxMC0wNi0yMy5jcnQwDQYJ
// SIG // KoZIhvcNAQELBQADggIBAJ1VffwqreEsH2cBMSRb4Z5y
// SIG // S/ypb+pcFLY+TkdkeLEGk5c9MTO1OdfCcTY/2mRsfNB1
// SIG // OW27DzHkwo/7bNGhlBgi7ulmZzpTTd2YurYeeNg2Lpyp
// SIG // glYAA7AFvonoaeC6Ce5732pvvinLbtg/SHUB2RjebYIM
// SIG // 9W0jVOR4U3UkV7ndn/OOPcbzaN9l9qRWqveVtihVJ9Ak
// SIG // vUCgvxm2EhIRXT0n4ECWOKz3+SmJw7wXsFSFQrP8DJ6L
// SIG // GYnn8AtqgcKBGUIZUnWKNsIdw2FzLixre24/LAl4FOmR
// SIG // sqlb30mjdAy87JGA0j3mSj5mO0+7hvoyGtmW9I/2kQH2
// SIG // zsZ0/fZMcm8Qq3UwxTSwethQ/gpY3UA8x1RtnWN0SCyx
// SIG // TkctwRQEcb9k+SS+c23Kjgm9swFXSVRk2XPXfx5bRAGO
// SIG // WhmRaw2fpCjcZxkoJLo4S5pu+yFUa2pFEUep8beuyOiJ
// SIG // Xk+d0tBMdrVXVAmxaQFEfnyhYWxz/gq77EFmPWn9y8FB
// SIG // SX5+k77L+DvktxW/tM4+pTFRhLy/AsGConsXHRWJjXD+
// SIG // 57XQKBqJC4822rpM+Zv/Cuk0+CQ1ZyvgDbjmjJnW4SLq
// SIG // 8CdCPSWU5nR0W2rRnj7tfqAxM328y+l7vzhwRNGQ8cir
// SIG // Ooo6CGJ/2XBjU02N7oJtpQUQwXEGahC0HVUzWLOhcGby
// SIG // oYIDUDCCAjgCAQEwgfmhgdGkgc4wgcsxCzAJBgNVBAYT
// SIG // AlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQH
// SIG // EwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29y
// SIG // cG9yYXRpb24xJTAjBgNVBAsTHE1pY3Jvc29mdCBBbWVy
// SIG // aWNhIE9wZXJhdGlvbnMxJzAlBgNVBAsTHm5TaGllbGQg
// SIG // VFNTIEVTTjo3RjAwLTA1RTAtRDk0NzElMCMGA1UEAxMc
// SIG // TWljcm9zb2Z0IFRpbWUtU3RhbXAgU2VydmljZaIjCgEB
// SIG // MAcGBSsOAwIaAxUAThIvkv2VRXusNSHd9ZuioHtupTSg
// SIG // gYMwgYCkfjB8MQswCQYDVQQGEwJVUzETMBEGA1UECBMK
// SIG // V2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwG
// SIG // A1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMSYwJAYD
// SIG // VQQDEx1NaWNyb3NvZnQgVGltZS1TdGFtcCBQQ0EgMjAx
// SIG // MDANBgkqhkiG9w0BAQsFAAIFAOlDM6wwIhgPMjAyNDAx
// SIG // MDYwMjA0MjhaGA8yMDI0MDEwNzAyMDQyOFowdzA9Bgor
// SIG // BgEEAYRZCgQBMS8wLTAKAgUA6UMzrAIBADAKAgEAAgIk
// SIG // OgIB/zAHAgEAAgIUJjAKAgUA6USFLAIBADA2BgorBgEE
// SIG // AYRZCgQCMSgwJjAMBgorBgEEAYRZCgMCoAowCAIBAAID
// SIG // B6EgoQowCAIBAAIDAYagMA0GCSqGSIb3DQEBCwUAA4IB
// SIG // AQAHuBoNAcULX2s8Qz3j3wnQtyXfwxdtCFb6nzb9KxWD
// SIG // MfOOiyvROSxXSWc1vBYQaSa3GWJrPa+1oUKrbkvJBA1o
// SIG // 5galiCAwwtQ12K3mZ1FFAk1hiyrfdv/Bo5gu05QHqmXc
// SIG // Vlhvstg/5dAYCDK8M9QS5WQyABwTzFNMF/5KGbjosUPG
// SIG // /cfBAiln98SvYR39tr6CRCD7gfnzv3vQF9n2dtTC8cuV
// SIG // lU+O/QcKkKL2VSj1iZbzR9QgmkEcCMFBBUTSXTlwT9/5
// SIG // dlbCc2MZ8qg6Wf/dCGk1Mcy/SKwOPcQWp/KxQXb9BGdp
// SIG // 3ktnJPyeN6RTfa4diFMjuik1GQJH7SpQIy3lMYIEDTCC
// SIG // BAkCAQEwgZMwfDELMAkGA1UEBhMCVVMxEzARBgNVBAgT
// SIG // Cldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAc
// SIG // BgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQG
// SIG // A1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIw
// SIG // MTACEzMAAAHVqQLPxafJ6VoAAQAAAdUwDQYJYIZIAWUD
// SIG // BAIBBQCgggFKMBoGCSqGSIb3DQEJAzENBgsqhkiG9w0B
// SIG // CRABBDAvBgkqhkiG9w0BCQQxIgQg3oC7TO6PXLZ7gVMe
// SIG // XKU7omQ9XZJ1I273fEWZRbb4NZQwgfoGCyqGSIb3DQEJ
// SIG // EAIvMYHqMIHnMIHkMIG9BCDZvyOGDNeuiTDOxCiHGL5X
// SIG // G69gh4ygtC1DpqWSGwbB/zCBmDCBgKR+MHwxCzAJBgNV
// SIG // BAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYD
// SIG // VQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQg
// SIG // Q29ycG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jvc29mdCBU
// SIG // aW1lLVN0YW1wIFBDQSAyMDEwAhMzAAAB1akCz8Wnyela
// SIG // AAEAAAHVMCIEIEtbOezXC+GuvHrl5/0ZGsc3M/hst8C1
// SIG // cGTHmqM3YtJZMA0GCSqGSIb3DQEBCwUABIICABm7NMNy
// SIG // KVaf4UPRG7krjgo+S+f2BWET1onMamw+B8SZGBZ8GFW+
// SIG // LovngHIAyE0vK3tLo63X6bzm0pXVMOt+17WBJc2JrP3Z
// SIG // AkfjiMJoAiPSx+s7vMkJrZ+EpSDtd/3a4H9ZRiauEvHp
// SIG // 5qxhGrSTPwWGUhq3elTAHB6EPYqn5dmg4x/5sO5koPvf
// SIG // kqHLTVFr7aw3NLHrsd8iVANcSs3rf5E8P2TxHDfDs/iI
// SIG // 0qFxUmiSAfXcP+VPhtJ3HKDB5q/GtoNvHNfaav93yuqg
// SIG // o/oDZclvFPTpwqNVGlpx8aUl+nBsNQaYTZ1/rA2kO2Cf
// SIG // 7JYJjd43LpR7TClVJWcAzhdwcVigqGf3ym/7Oj5Ls4wb
// SIG // XfbOITV4V+Xh4RLScpn0+IkjTUYLOvj4UWRoVn3HcU5Z
// SIG // myMK3tgEZxrzoQ8uVcr8uAvLavR6loahFMmReLWCdR72
// SIG // BnyNiTXNxCcXXFHmu6tw5JROg3iwgeZMC6cTThqoWIUD
// SIG // 1Nsx07XCYDQu6B+v6Lgc0XsBqLKVLUP42HeZGKhZoGA6
// SIG // CujkMhYbLUHyXGFpA6aVAY4/KA3Tz8Od4U45NonyuftA
// SIG // 1L1EQtiNTpTcztKn1+cCxG2jAyYUm/8gWTP1YyXKALaB
// SIG // H6ljBML4PiNNii/EI4uZuw3In0Qr9gAYBW6MSvu0iJXy
// SIG // QgT/w80A/0ciC/3L
// SIG // End signature block
